extmod/modbluetooth.c

```diff
@@ -1168,6 +1171,45 @@ STATIC mp_obj_t invoke_irq_handler_run(uint16_t event,
         data_tuple->items[data_tuple->len++] = MP_OBJ_FROM_PTR(uuid);
     }
     #endif
+
+    #if MICROPY_PY_BLUETOOTH_USE_GATTC_EVENT_DATA_REASSEMBLY
+    void *buf_to_free = NULL;
+    size_t buf_to_free_len = 0;
+    if (event == MP_BLUETOOTH_IRQ_GATTC_NOTIFY || event == MP_BLUETOOTH_IRQ_GATTC_INDICATE || event == MP_BLUETOOTH_IRQ_GATTC_READ_RESULT) {
+        assert(n_data >= 1);
+
+        const uint8_t *combined_data;
+        size_t total_len;
+        if (n_data > 1) {
+            // Fragmented buffer, need to combine into a new heap-allocated buffer
+            // in order to pass to Python.
+            total_len = 0;
+            for (size_t i = 0; i < n_data; ++i) {
+                total_len += data_len[i];
+            }
+            uint8_t *buf = m_new(uint8_t, total_len);
+            uint8_t *p = buf;
+            for (size_t i = 0; i < n_data; ++i) {
+                memcpy(p, data[i], data_len[i]);
+                p += data_len[i];
+            }
+            combined_data = buf;
```

Comment:
Can you remove the combined_data variable.

Then on this line, instead:
```
data[0] = buf;
data_len = 1;
```

Remove the else below, and the following lines til the `}` above the `#endif`.

Then let the `i<n_data` loop run as normal.

Reply:
Ah, OK.  And that's only possible because in `gattc_on_data_available` (the only user of this part of the code) allocates these `data` and `data_len` arrays itself and doesn't care if they are written to.  I'll add a comment to that effect here.